File: 1171. Remove Zero Sum Consecutive Nodes from Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeZeroSumSublists(ListNode head) {
        if (head == null) return null;
        ListNode pointer = head;
        int sum = 0;
        while (pointer != null) {
            sum += pointer.val;
            pointer = pointer.next;
            if (sum == 0) return removeZeroSumSublists(pointer);
        }
        head.next = removeZeroSumSublists(head.next);
        return head;
    }
}

File: go/1171. Remove Zero Sum Consecutive Nodes from Linked List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeZeroSumSublists(head *ListNode) *ListNode {
    if head == nil {
        return nil
    }
    sum := 0
    pointer := head
    for pointer != nil {
        sum += pointer.Val
        pointer = pointer.Next
        if sum == 0 {
            return removeZeroSumSublists(pointer)
        }
    }
    head.Next = removeZeroSumSublists(head.Next)
    return head
}

